File: sdk/typescript/src/turn.ts
import type { Block, LogEvent, RunEvent, TurnState } from "./types.ts";

const TERMINAL_TURN_STATES = new Set(["done", "failed", "interrupted"]);

/**
 * Folds the log feed into one turn's answer.
 *
 * The feed carries every turn of the conversation and every stream of each
 * turn, so "what did the agent just say" is a filtering problem, not a
 * concatenation one. This class is the filter:
 *
 *   - a `stage`/`turn` event opens and closes the turn we are following, and
 *     is the only thing that says how it ended;
 *   - `output` events carry server-parsed `blocks`; only `text` is the answer,
 *     `tool_use` is noise worth naming, `thinking` is neither;
 *   - text that follows a tool call is a new message, so it gets a paragraph
 *     break — the rule that stops a transcript reading as one run-on sentence.
 *
 * The joining rules are ported from the Hermes plugin, which learned them
 * against real runtimes: ACP streams one message as chunks that join with
 * nothing, while a legacy stdout row is a whole message and joins as a
 * paragraph.
 */
export class TurnFollower {
  readonly turnNumber: number;
  turnId: string | null = null;
  started = false;
  state: TurnState | null = null;
  exitCode: number | null = null;
  reason: string | null = null;

  private readonly chunks: string[] = [];
  private readonly tools: string[] = [];
  private breakBeforeText = false;

  constructor(turnNumber: number, turnId: string | null = null) {
    this.turnNumber = turnNumber;
    this.turnId = turnId;
  }

  get text(): string {
    return this.chunks.join("").trim();
  }

  get toolsUsed(): string[] {
    return [...this.tools];
  }

  get finished(): boolean {
    return this.state !== null;
  }

  /** Fold one event in, and report what a streaming caller should be told. */
  apply(event: LogEvent): RunEvent[] {
    if (event.kind === "stage") return this.applyStage(event);
    if (event.kind !== "output") return [];
    return this.applyOutput(event);
  }

  private applyStage(event: LogEvent): RunEvent[] {
    if (event.stage !== "turn") return [];
    const meta = parseJson(event.data) ?? {};
    if (!this.matchesTurn(meta)) return [];

    if (event.state === "started") {
      this.started = true;
      this.turnId = asString(meta.turn_id) ?? this.turnId;
      return [{ type: "turn-start", turnNumber: this.turnNumber, turnId: this.turnId }];
    }

    if (event.state && TERMINAL_TURN_STATES.has(event.state)) {
      this.state = event.state as TurnState;
      this.turnId = this.turnId ?? asString(meta.turn_id);
      if (typeof meta.exit_code === "number") this.exitCode = meta.exit_code;
      const reason = asString(meta.reason) ?? asString(meta.stop_reason);
      if (reason) this.reason = reason;
      return [
        { type: "turn-end", state: this.state, exitCode: this.exitCode, reason: this.reason },
      ];
    }

    return [];
  }

  private matchesTurn(meta: Record<string, unknown>): boolean {
    if (this.turnId && asString(meta.turn_id) === this.turnId) return true;
    return meta.turn_number === this.turnNumber;
  }

  private applyOutput(event: LogEvent): RunEvent[] {
    const turnId = asString(event.turn_id);
    // A different turn's output — history, or a turn someone else started.
    if (this.turnId && turnId && turnId !== this.turnId) return [];
    // Output from before our turn opened is the tail of an older one.
    if (!this.started && !this.turnId) return [];

    const acp = event.stream === "acp";
    const out: RunEvent[] = [];

    for (const block of event.blocks ?? []) {
      out.push({ type: "block", block, event });
      out.push(...this.applyBlock(block, acp));
    }
    return out;
  }

  private applyBlock(block: Block, acp: boolean): RunEvent[] {
    const body = block.body ?? "";

    if (block.kind === "text") {
      if (!body) return [];
      const prefix = this.paragraphBreak(acp);
      if (prefix) this.chunks.push(prefix);
      this.chunks.push(body);
      this.breakBeforeText = false;
      return [{ type: "text", text: prefix + body }];
    }

    if (block.kind === "thinking") {
      return body ? [{ type: "thinking", text: body }] : [];
    }

    // `raw` and `init` are transport bookkeeping: not output, and not a break.
    if (block.kind === "raw" || block.kind === "init") return [];

    this.breakBeforeText = true;

    if (block.kind === "tool_use") {
      const name = asString(block.name);
      if (!name) return [];
      if (!this.tools.includes(name)) this.tools.push(name);
      return [{ type: "tool", name, block }];
    }

    // A `result` block is the answer only when the runtime said nothing else.
    if (block.kind === "result" && !this.chunks.length && body) {
      this.chunks.push(body);
      return [{ type: "text", text: body }];
    }

    if (block.kind === "error" && body) {
      const text = `\n[error] ${body}\n`;
      this.chunks.push(text);
      return [{ type: "text", text }];
    }

    return [];
  }

  /**
   * ACP chunks are pieces of one message and join with nothing; anything after
   * a tool call is a new message. A legacy row is a whole message either way.
   */
  private paragraphBreak(acp: boolean): string {
    if (!this.chunks.length) return "";
    if (acp && !this.breakBeforeText) return "";
    const last = this.chunks[this.chunks.length - 1] ?? "";
    return last.endsWith("\n") ? "" : "\n\n";
  }
}

function parseJson(raw: unknown): Record<string, unknown> | null {
  if (raw && typeof raw === "object") return raw as Record<string, unknown>;
  if (typeof raw !== "string" || !raw.trim()) return null;
  try {
    const parsed = JSON.parse(raw);
    return parsed && typeof parsed === "object" ? (parsed as Record<string, unknown>) : null;
  } catch {
    return null;
  }
}

function asString(value: unknown): string | null {
  return typeof value === "string" && value ? value : null;
}

File: sdk/python/src/fountain/turn.py
"""Fold a conversation's event feed into one turn."""

import json
from typing import Any, Dict, List, Optional

_TERMINAL_STATES = {"done", "failed", "interrupted"}


class TurnFollower:
    def __init__(self, turn_number: int, turn_id: Optional[str] = None) -> None:
        self.turn_number = turn_number
        self.turn_id = turn_id
        self.started = False
        self.state: Optional[str] = None
        self.exit_code: Optional[int] = None
        self.reason: Optional[str] = None
        self._chunks: List[str] = []
        self._tools: List[str] = []
        self._break_before_text = False

    @property
    def text(self) -> str:
        return "".join(self._chunks).strip()

    @property
    def tools_used(self) -> List[str]:
        return list(self._tools)

    @property
    def finished(self) -> bool:
        return self.state is not None

    def apply(self, event: Dict[str, Any]) -> List[Dict[str, Any]]:
        if event.get("kind") == "stage":
            return self._apply_stage(event)
        if event.get("kind") == "output":
            return self._apply_output(event)
        return []

    def _apply_stage(self, event: Dict[str, Any]) -> List[Dict[str, Any]]:
        if event.get("stage") != "turn":
            return []
        meta = _object(event.get("data"))
        if not self._matches(meta):
            return []
        state = event.get("state")
        if state == "started":
            self.started = True
            self.turn_id = _string(meta.get("turn_id")) or self.turn_id
            return [
                {
                    "type": "turn-start",
                    "turn_number": self.turn_number,
                    "turn_id": self.turn_id,
                }
            ]
        if state in _TERMINAL_STATES:
            self.state = state
            self.turn_id = self.turn_id or _string(meta.get("turn_id"))
            if isinstance(meta.get("exit_code"), int):
                self.exit_code = meta["exit_code"]
            self.reason = (
                _string(meta.get("reason"))
                or _string(meta.get("stop_reason"))
                or self.reason
            )
            return [
                {
                    "type": "turn-end",
                    "state": self.state,
                    "exit_code": self.exit_code,
                    "reason": self.reason,
                }
            ]
        return []

    def _matches(self, meta: Dict[str, Any]) -> bool:
        if self.turn_id and _string(meta.get("turn_id")) == self.turn_id:
            return True
        return meta.get("turn_number") == self.turn_number

    def _apply_output(self, event: Dict[str, Any]) -> List[Dict[str, Any]]:
        event_turn_id = _string(event.get("turn_id"))
        if self.turn_id and event_turn_id and event_turn_id != self.turn_id:
            return []
        if not self.started and not self.turn_id:
            return []
        output: List[Dict[str, Any]] = []
        for block in event.get("blocks") or []:
            if not isinstance(block, dict):
                continue
            output.append({"type": "block", "block": block, "event": event})
            output.extend(self._apply_block(block, event.get("stream") == "acp"))
        return output

    def _apply_block(self, block: Dict[str, Any], acp: bool) -> List[Dict[str, Any]]:
        kind = block.get("kind")
        body = block.get("body") if isinstance(block.get("body"), str) else ""
        if kind == "text":
            if not body:
                return []
            prefix = self._paragraph_break(acp)
            self._chunks.extend((prefix, body))
            self._break_before_text = False
            return [{"type": "text", "text": prefix + body}]
        if kind == "thinking":
            return [{"type": "thinking", "text": body}] if body else []
        if kind in ("raw", "init"):
            return []
        self._break_before_text = True
        if kind == "permission_request":
            request = _permission_request(block)
            return (
                [{"type": "permission", "request": request, "block": block}]
                if request
                else []
            )
        if kind == "tool_use":
            name = _string(block.get("name"))
            if not name:
                return []
            if name not in self._tools:
                self._tools.append(name)
            return [{"type": "tool", "name": name, "block": block}]
        if kind == "result" and not self._chunks and body:
            self._chunks.append(body)
            return [{"type": "text", "text": body}]
        if kind == "error" and body:
            text = "\n[error] %s\n" % body
            self._chunks.append(text)
            return [{"type": "text", "text": text}]
        return []

    def _paragraph_break(self, acp: bool) -> str:
        if not self._chunks or (acp and not self._break_before_text):
            return ""
        return "" if self._chunks[-1].endswith("\n") else "\n\n"


def _permission_request(block: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    request_id = _string(block.get("request_id"))
    if not request_id:
        return None
    options = []
    for raw in block.get("options") or []:
        if not isinstance(raw, dict):
            continue
        option_id = _string(raw.get("optionId")) or _string(raw.get("option_id"))
        if option_id:
            option = dict(raw)
            option["option_id"] = option_id
            options.append(option)
    if not options:
        return None
    return {
        "request_id": request_id,
        "summary": _string(block.get("summary")) or _string(block.get("body")),
        "tool_name": _string(block.get("name")),
        "tool_id": _string(block.get("tool_id")),
        "options": options,
    }


def _object(value: Any) -> Dict[str, Any]:
    if isinstance(value, dict):
        return value
    if not isinstance(value, str) or not value.strip():
        return {}
    try:
        parsed = json.loads(value)
    except json.JSONDecodeError:
        return {}
    return parsed if isinstance(parsed, dict) else {}


def _string(value: Any) -> Optional[str]:
    return value if isinstance(value, str) and value else None
